gcc/rust/backend/rust-compile-proc-macro.cc

```diff
@@ -0,0 +1,393 @@
+// - Store the address of the function
+tree
+bang_proc_macro (tree macro)
+{
+  // Attribute and bang proc macros have the same structure, they can be
+  // initialized with the same code.
+  return attribute_proc_macro (macro);
+}
+
+// Initialize a proc macro structure from a given payload tree
+tree
+proc_macro (tree payload, tree proc_macro_type, ProcMacro::ProcmacroTag tag)
+{
+  auto discriminant = static_cast<int> (tag);
+
+  tree macro_tag = build_int_cst (short_unsigned_type_node, discriminant);
+
+  tree payload_union
+    = Backend::constructor_expression (build::proc_macro_payload (), false,
+				       {payload},
+				       discriminant /* Union: member index */,
+				       BUILTINS_LOCATION);
+
+  return Backend::constructor_expression (proc_macro_type,
+					  false /* invariant */,
+					  {macro_tag, payload_union},
+					  -1 /* Structure: No index */,
+					  BUILTINS_LOCATION);
+}
+
+tree
+proc_macro_array (Context *ctx, tree proc_macro_buffer_type,
+		  tree proc_macro_type)
+{
+  std::vector<unsigned long> indexes;
+  std::vector<tree> ctors;
+  size_t index = 0;
+  for (auto &macro : ctx->get_derive_proc_macros ())
+    {
+      tree derive = derive_proc_macro (ctx, macro);
+      ctors.push_back (proc_macro (derive, proc_macro_type,
+				   ProcMacro::ProcmacroTag::CUSTOM_DERIVE));
+      indexes.push_back (index);
+      index++;
+    }
+  for (auto &macro : ctx->get_attribute_proc_macros ())
+    {
+      tree attr = attribute_proc_macro (macro);
+
+      ctors.push_back (
+	proc_macro (attr, proc_macro_type, ProcMacro::ProcmacroTag::ATTR));
+      indexes.push_back (index);
+      index++;
+    }
+  for (auto &macro : ctx->get_bang_proc_macros ())
+    {
+      tree bang = bang_proc_macro (macro);
+
+      ctors.push_back (
+	proc_macro (bang, proc_macro_type, ProcMacro::ProcmacroTag::BANG));
+      indexes.push_back (index);
+      index++;
+    }
+
+  auto length = build_int_cst (unsigned_type_node, ctors.size ());
+  auto array = Backend::array_constructor_expression (
+    build_array_type_nelts (proc_macro_type, ctors.size ()), indexes, ctors,
+    BUILTINS_LOCATION);
+  return Backend::constructor_expression (proc_macro_buffer_type,
+					  false /* invariant */,
+					  {length, array},
+					  -1 /* Structure: No index */,
+					  BUILTINS_LOCATION);
+}
+} // namespace init
+
+} // namespace
+
+// Gather procedural macros and generate the metadata as well as the entrypoint
+// for a procedural macro crate.
+void
+CompileCrate::add_proc_macro_symbols ()
+{
+  auto total_macros = ctx->get_attribute_proc_macros ().size ()
+		      + ctx->get_bang_proc_macros ().size ()
+		      + ctx->get_derive_proc_macros ().size ();
+
+  tree pm_type = build::proc_macro ();
+  tree pm_buffer_type = build::proc_macro_buffer (pm_type, total_macros);
+  tree entrypoint_type = build::entrypoint (pm_buffer_type);
+
+  std::string decl_symbol_name = generate_proc_macro_decls_symbol (
+    0 /* FIXME: Change to stable crate id */);
```

Comment:
You should be able to get mappings->get_current_crate for that.

Reply:
Are your sure we're talking about the same crate id ? IIRC the stable crate id is made from the hash of the crate's metadata. I opened #2696 for this a while ago, but since you wanted to rework the metadata format I left it as is while we wait for this rework.